### ISSUE TITLE
calc performance with large selections - copy/paste

### ISSUE THREAD

**Issue by mmeeks** [CollaboraOnline/online#6893]
Editing just now on staging with a larger group we got a number of problems with calc performance that should be improved.

![perf-23 05-calc](https://github.com/CollaboraOnline/online/assets/122848/1ce72623-30e9-4f0c-947b-b7440b6ec400)

Oddly we were getting large memory consumption for a smallish spreadsheet too - perhaps related to clipboads:

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
32675 cool      20   0 20.003t 2.904g 343088 R 103.0 9.263   6:31.37 kitbroker_012

which is really unexpected.

**Comment by mmeeks** [comment]
Lets get the document from Pedro & can you look at this  @eszkadev ?

**Comment by eszkadev** [comment]
I cannot reproduce case where getTextSelection takes significant piece of a profile. Tested with large selections and others modifying the content.

I think in this ticket we see long-term profile where most of the operations were selection changes, people were clicking here and there without typing or any other actions - so that is why we see it bigger.

In my profiles where editing happens bigger problem is rendering of the tiles or full column selection.

**Comment by mmeeks** [comment]
Hmm - I strongly suspect the memory blow-out and swapping which matches that profile and the very significant memory use can only come from that. I imagine that it is specific to Pedro's document which oddly has all 16k columns allocated for me too (causing all manner of problems =)

**Comment by caolanm** [comment]
wrt:

PID USER PR NI VIRT RES SHR S %CPU %MEM TIME+ COMMAND
32675 cool 20 0 20.003t 2.904g 343088 R 103.0 9.263 6:31.37 kitbroker_012

the startling virt num indicates an asan build, for a fresh blank spreadsheet under asan I get

PID USER PR NI VIRT RES SHR S %CPU %MEM TIME+ COMMAND
11532 cool 20 0 20.003t 2.763g 318000 S 0.000 8.816 0:05.36 kitbroker_003

before we do anything.

**Comment by pedropintosilva** [comment]
I think this was it :
* https://staging.eu.collaboraonline.com/nextcloud/index.php/s/kpKMewkjTreSDM4
    * [COOLWeekly127-latest.ods](https://github.com/CollaboraOnline/online/files/12319669/COOLWeekly127-latest.ods)
* Another version of the document [COOLWeekly127.ods](https://github.com/CollaboraOnline/online/files/12319670/COOLWeekly127.ods)

**Comment by caolanm** [comment]
This weeks calc test looks like:
![perf-calc](https://github.com/CollaboraOnline/online/assets/833656/a2bf878e-576e-4404-86f1-81a64ff33054)

**Comment by caolanm** [comment]
I also see:
 * With one window on a tab X with another window on tab Y. Editing in the first view on tab X and then moving cursor to another cells "kills the editview" and the other views are invalidated for the same area, but on whatever tab they happen to be looking at, so tab Y gets invalidated in the same area. https://gerrit.libreoffice.org/c/core/+/155816 for a little chain of commits on that thought. Blinkenlights still flash for the invalidated area on the other sheet but tile-invalidation count doesn't change.

**Comment by caolanm** [comment]
This weeks flamegraph for calc ![perf-staging-calc-2023-08-24](https://github.com/CollaboraOnline/online/assets/833656/e3058b6f-c3d6-46a0-87c1-8fb83940043d)

**Comment by mmeeks** [comment]
We should really look at that view switching cost: doc_setView should be the moral equivalent of a single pointer change it's strobed at very high frequency =) Instead in this profile it unzips, and de-compresses a (small) PNG - worth checking if it does it every time - but we should be able to save some 30%+ of our paintPartTile cost there I think :-)

It is also interesting how little the thread pool does to help with the compression and delta generation; I'd like to hope this is because calc is good at only invalidating one tile as we type into a cell - so that is done in the main-thread but ... quite possibly there's a snafu there too.

**Comment by caolanm** [comment]
This weeks flamegraph for calc

![perf-8556](https://github.com/CollaboraOnline/online/assets/833656/6345ee72-2a06-402e-a12c-72bd8d49df8d)

**Comment by caolanm** [comment]
This weeks calc's perf flamegraph

![perf-22435](https://github.com/CollaboraOnline/online/assets/833656/79c39c7c-a482-4d3c-b9bf-6cfabfa56f2e)

https://github.com/CollaboraOnline/online/issues/7337 seems to have worked and at least that has shrunk

**Comment by caolanm** [comment]
Clearly the Calc navigator is out of control in multiview documents wrt doc_setView triggering such much updates. I'll look into that.

**Comment by caolanm** [comment]
Two framegraphs this week: The first for the mostly inactive session until the crash which is visible in profiling
![perf-19099](https://github.com/CollaboraOnline/online/assets/833656/03852ae9-0a55-4083-92c6-642cea57c77c)

The second one on the restart to end:
![perf-21240](https://github.com/CollaboraOnline/online/assets/833656/f8946ec3-74da-4d50-91fc-518d6c14639c)

**Comment by mmeeks** [comment]
Interesting: the mouse event processing seems to have an un-wanted inferior main-loop from chart::ChartControlller::executeDlg_ObjectProperties_withoutUndoGuard - which needs to be async'd:

![image](https://github.com/CollaboraOnline/online/assets/122848/c488ba47-0194-47f8-ab2f-014b840a2d59)

And of course the amount of work done in doc_setView - to update the UI / notebookbar is visible everywhere; I'll file a separate ticket for that.

**Comment by mmeeks** [comment]
@grandinj This ticket is probably interesting for you with its weekly feed of profiling porn :-)

**Comment by grandinj** [comment]
related to @caolanm flamegraphs at comment https://github.com/CollaboraOnline/online/issues/6893#issuecomment-1759411348_
I have created https://github.com/CollaboraOnline/online/pull/7430 to fix a small perf problem.

**Comment by caolanm** [comment]
This weeks calc flamegraph. We're skewed by the threaded formula thing but expanding kitbroker_006 to look at that alone at least we don't have the ridiculous constant creation of a WeldedNotebookbar of last week noted as https://github.com/CollaboraOnline/online/issues/7428 though still too much going on in doc_setView

![perf-14321](https://github.com/CollaboraOnline/online/assets/833656/d1eb6343-520f-42c7-b038-8ae3e8d856d0)

**Comment by caolanm** [comment]
pre-crash calc profile:

![perf-19984](https://github.com/CollaboraOnline/online/assets/833656/fb0d5ad1-335c-413e-98fc-66559bd033c6)

looks like:

a) MergeCells is not an async dialog
b) crash is visible as coming from ScDocShell::MoveTable

post-crash calc profile:

![perf-21534](https://github.com/CollaboraOnline/online/assets/833656/4cf8804b-4f60-4ac5-90e2-28c4e59f2867)

The problem of https://github.com/CollaboraOnline/online/issues/7428 is not totally gone, but the cost of it is less now without the notebookbar torn down and recreated for dark mode switching so there seems likely a relation to that in the earlier profiling

**Comment by caolanm** [comment]
![perf-15373](https://github.com/CollaboraOnline/online/assets/833656/ef3d86b1-23ea-4e57-ad15-1f9b12df6bdd)

Got an apparent modal SvxCharacterMap dialog, some hefty CalendarWrapper work and spell check work.

**Comment by grandinj** [comment]
(1) I note that hunspell is spending a lot of time just calling clock(), which seems odd. 
Perhaps it should be using clock_gettime(CLOCK_MONOTONIC) instead, to use the fancy mapping trick that Linux does and avoid a syscall.

(2) we seem to be spending a lot of time just figuring out which dictionary to use, not even doing actual dictionary loading or actual spell-checking. Surely that should be cached somewhere?

**Comment by grandinj** [comment]
possible fix for the hunspell issue is here: https://gerrit.libreoffice.org/c/core/+/158809

**Comment by caolanm** [comment]
https://gerrit.libreoffice.org/c/core/+/158830 for the i18npool::Calendar_jewish::mapFromGregorian issue, its nearly always 1999-12-31 so we can just add a fast path for that

**Comment by caolanm** [comment]
This weeks calc profile:
![perf-6686](https://github.com/CollaboraOnline/online/assets/833656/d6d856d2-f8f2-4f7e-9011-c292262058bc)

**Comment by caolanm** [comment]
For reference in today's bespoke tests:
First session:
![perf-27903](https://github.com/CollaboraOnline/online/assets/833656/5e9634bf-953f-4c1a-bcaf-ca3dda1bec5f)

@hcvcastro  The ScDBData::ExtendBackColorArea is a pretty recent new method I believe, since
https://cgit.freedesktop.org/libreoffice/core/commit/?id=768433f07873eb608837630f85e7e1b375239fca
Seems to take up quite a bit of time.

Second session:
![perf-29178](https://github.com/CollaboraOnline/online/assets/833656/285e2438-1a28-4794-8e18-cbefb0ce9323)

**Comment by caolanm** [comment]
![perf-18345](https://github.com/CollaboraOnline/online/assets/833656/fd1ef8d4-749b-4d96-a427-534e6ff0b4e6)

This week

**Comment by grandinj** [comment]
IRC log in case someone wants to look at the ScPositionHelper thing, I dont know what is going on there:

<mmeeks> void ScGridWindow::PaintTile( VirtualDevice& rDevice, - doing a big/slow __memset is unexpected - particularly that the size is larger than the pixel-size of the tile - which is quite large ...
* mmeeks wonders if some big hash-table or STL type is fiddled with there.
<mmeeks> @noel_grandin: seems there is a ton of work done on HandleMouseEvent =) 12.5% of total time handling mouse events seems extraordinary =)
<mmeeks> a greath swath of GetRowHeight going on there - from the drawing-layer =)
<noel_grandin> weird, if we are hitting GetRowHeight that means that the ScPositionHelper cache is not doing its job

**Comment by caolanm** [comment]
![perf-15753](https://github.com/CollaboraOnline/online/assets/833656/67705adc-df42-40b7-a21e-0016b7c7229c)

This weeks. No crashes, so 1h. ScTabViewShell::ExecuteTable has some non-async dialogs still

**Comment by grandinj** [comment]
The calloc in the graph is now at 5% and is unnecessary - no need to memset() there, the data will be completely override during the delta generation process.

**Comment by mmeeks** [comment]
If we can safely rid ourselves of the calloc that's great of course =)

**Comment by caolanm** [comment]
![perf-10543](https://github.com/CollaboraOnline/online/assets/833656/616c943e-8cc6-4b95-92ef-69cf72ec57ee)

Todays little bespoke test, I wonder about ExtendBackColorArea, I only clicked that AutoFilter button two or three times, but it takes up quite a wide part

**Comment by caolanm** [comment]
Today's "share" deployment profile

![perf-71519](https://github.com/CollaboraOnline/online/assets/833656/f615306c-5234-4ce1-92f8-e9b52602285a)

**Comment by caolanm** [comment]
Not sure if https://gerrit.libreoffice.org/c/core/+/160784 is a legitimate optimization for the ScColContainer::resize seen from ScDocument::GetDocStat

**Comment by caolanm** [comment]
Todays profile as far as crash

 ![perf-170021](https://github.com/CollaboraOnline/online/assets/833656/f5211b4c-38c5-4b03-8a61-6dbc42f77d97)

**Comment by caolanm** [comment]
After: ![perf-172885](https://github.com/CollaboraOnline/online/assets/833656/004f09ae-7a9f-42ea-ae95-988f5f252579)

Crash in 1st profile can be seen as chart::ChartModel::lockController

**Comment by caolanm** [comment]
![hotspot](https://github.com/CollaboraOnline/online/assets/833656/bec538c0-dd56-4190-b7ec-0a94fe2c5324)

![hotspot2](https://github.com/CollaboraOnline/online/assets/833656/d804c690-6272-41e3-b53b-a46baf57e20d)

FWIW SalLayoutGlyphsCache::self has 20000 as the max size of this std::list. While I would expect this to be c++11 std::list I wonder if because these binaries are built with "Red Hat Developer Toolset 10" given /opt/rh/devtoolset-10 in the binaries if they end up built in some C++03 compatible ABI mode.

**Comment by mmeeks** [comment]
Wow, std::list is an horribly efficient pointer-chasing nightmare surely that shouldn't be used on modern machines =) @grandinj have we looked std::vector there or something else more efficient; I assume it was never intended for 20k items (?) or our access pattern. Seems rather likely that we could have a monotonic 'generation' id on each item, and a high/low watermark for scanning & pruning and removing the oldest things (?) anyhow a fertile place at 25% of the cost of calc to optimize I think =)

**Comment by caolanm** [comment]
![perf-441885](https://github.com/CollaboraOnline/online/assets/833656/25c6a517-65c4-4c89-bec1-2dea5fb90e54)

Probably not particularly useful as only the tail end of session with 3 participants, but I'll attach anyway.

**Comment by mmeeks** [comment]
Well, half of the time in the DeltaGenerator again is perhaps a good sign that nothing more silly is going on (or we get way too many invalidations ;-) - although its really quite surprising the performance of that compared to the (much more complex) rendering. I suspect we should track invalidations vs. tiles and copy across unmodified horizontal lines wherever we can to improve life here without RLE'ing them again =)

**Comment by caolanm** [comment]
Todays calc profile

![perf-680008](https://github.com/CollaboraOnline/online/assets/833656/d8727387-178b-4679-ac48-cd4991aa5f1f)

**Comment by grandinj** [comment]
We are spending an awful lot of time in memset when doing ScGridWindow::PaintTile. 
Looking at the code, I think that this

    ScTableInfo aTabInfo(nEndRow + 3);

is likely to be the problem, especially when scrolling down in documents with lots of rows, because it will an array of RowInfo objects of size nEndRow + 3.

We can probably split that into two structures - one set of RowInfo objects for the rows before the tile starts, and one set of rows for the tile itself. For the rows before the tile starts, we likely need much less info.

**Comment by mmeeks** [comment]
Wow - that does look silly, if nEndRow is one million with a large sheet area (nTiledRenderingAreaEndRow) - or a single bad cell at the bottom, that's quite an expense if you're looking at the bottom of the sheet; Surely we only really need row information close to hand for the range of cells we are rendering for, and not this =)

**Comment by caolanm** [comment]
![perf-128979](https://github.com/CollaboraOnline/online/assets/833656/efbc630a-65af-4a4f-ae5e-b1daf60ceab2)

Todays calc which effectively hung, with a different spreadsheet than we typically use

**Comment by caolanm** [comment]
![perf-107896](https://github.com/CollaboraOnline/online/assets/833656/75e43dcb-918a-4d82-aa45-b6e7734f2d27)

This weeks calc session including crash

**Comment by mmeeks** [comment]
Interesting protocol thrashing here: https://github.com/CollaboraOnline/online/issues/8159 from some of this session.

**Comment by Tex2002ans** [comment]
For ease of finding later, I'll add a link to these Calc flamegraphs too:

- https://github.com/CollaboraOnline/online/issues/7330
- https://github.com/CollaboraOnline/online/issues/7332

**Comment by caolanm** [comment]
![perf-183388](https://github.com/CollaboraOnline/online/assets/833656/c9255e06-4e7e-4a52-bab1-8fcb3ef9298e)

first part of the calc session

**Comment by mmeeks** [comment]
The trace has lots of comedic waste in it: 
[calc-trace-2024-02-08.txt.gz](https://github.com/CollaboraOnline/online/files/14208723/calc-trace-2024-02-08.txt.gz)
I'll file some tickets for that.

**Comment by caolanm** [comment]
![perf-186739](https://github.com/CollaboraOnline/online/assets/833656/cf84e0c4-9a45-4cfd-9d78-b9ad6209f1eb)

second half of session

**Comment by caolanm** [comment]
![perf-295409](https://github.com/CollaboraOnline/online/assets/833656/5e32e38b-658b-4580-95b9-48d8395515b2)

todays ad-hoc call

**Comment by mmeeks** [comment]
ScXMLExport::collectAutoStyles looks like something that does far too much work - fouled up by the UNO interface there too I guess; 11% of the profile and ~half of saving time; @grandinj did you look at that call-site ever? =)

**Comment by caolanm** [comment]
various profiles of todays calc session
![perf-357585](https://github.com/CollaboraOnline/online/assets/833656/9140ce64-c2ff-46d0-80e1-8854b244859c)
![perf-357796](https://github.com/CollaboraOnline/online/assets/833656/f4b10503-0df5-457e-aeba-e324b79f9fc6)
![perf-359184](https://github.com/CollaboraOnline/online/assets/833656/44fa0f04-4d27-43e9-9a71-802ca2ece04d)

**Comment by mmeeks** [comment]
Filed https://github.com/CollaboraOnline/online/issues/8275 - for the expensive theme color notification thrash, and https://github.com/CollaboraOnline/online/issues/8276 for yet more note export waste.

**Comment by caolanm** [comment]
![perf-459395](https://github.com/CollaboraOnline/online/assets/833656/56903a71-71ca-42b2-8426-fc14d4803b8b)
![perf-460766](https://github.com/CollaboraOnline/online/assets/833656/682a43b5-85c8-42a8-ba62-c56091e73957)

flamegraphs from todays ad-hoc calc test, strongly dominated by chart related activity

**Comment by caolanm** [comment]
![perf-529829](https://github.com/CollaboraOnline/online/assets/833656/bf68e7b3-9113-41ae-bf9e-6c0760d2cc82)

todays calc profile

**Comment by caolanm** [comment]
![perf-52491](https://github.com/CollaboraOnline/online/assets/833656/1f01a246-b97a-4f0e-afd3-9e993c16f557)

todays ad-hoc calc session

**Comment by caolanm** [comment]
What we ended up with for ScTableInfo was to make it a hint anywhere we "traditionally" defaulted to 1024 rows before https://gerrit.libreoffice.org/c/core/+/162277 so that after https://gerrit.libreoffice.org/c/core/+/162875 we maxed at 1024 if 1024 was what we used before.

But for cases like https://gerrit.libreoffice.org/c/core/+/162277/2/sc/source/ui/view/gridwin4.cxx#b1667 where we had an explicit size passed in then we continue to use an explicit requested size (i.e. hint is false). So I think this case is the above line where we "always" used an explicit large range (and this is drawing a range with all rows hidden from some point downwards I think). Not sure if there is an obvious calculation for what the max number of visible rows can be drawn there is, or if its safe to stick some reasonable by largish max there too for that explicit rows requested case.

**Comment by kohei-us** [comment]
> Hmm; the previous profile is back to that huge allocation for ScTableInfo::ScTableInfo - which is odd; I thought we clobbered that, and ScBoundsProvider::Compute seems very hungry. This last one - is still plagued by the ScColumn::HasCellNotes expense @kohei-us wondered if you've seen any of these ? =)

I haven't profiled Calc in years!  But maybe this will help: https://gerrit.libreoffice.org/c/core/+/163996

**Comment by caolanm** [comment]
![perf-135921](https://github.com/CollaboraOnline/online/assets/833656/2180e413-64f9-4001-a909-b8b739a98228)

todays calc profile (dominated by chart drawing)

**Comment by grandinj** [comment]
On 2/29/2024 2:09 PM, Caolán McNamara wrote:
> todays calc profile (dominated by chart drawing)
> 

There is a still a chunk of memset in RenderTiles::doRender that looks unnecessary, but last time I tried to get rid of 
it, unit tests broke in weird places.

**Comment by grandinj** [comment]
> What we ended up with for ScTableInfo was to make it a hint anywhere we "traditionally" defaulted to 1024 rows before https://gerrit.libreoffice.org/c/core/+/162277 so that after https://gerrit.libreoffice.org/c/core/+/162875 we maxed at 1024 if 1024 was what we used before.
> 

Possibly we should ScTableInfo use a std::vector instead of a fixed array length, and adjust size as needed. Then we can pass in a hint like I did initially (and the hint will be wrong if there are hidden rows), but it will just adjust size as needed.
That will be slower for some cases (with hidden rows), but faster for all other cases.

**Comment by caolanm** [comment]
![perf-220007](https://github.com/CollaboraOnline/online/assets/833656/20545b4a-32da-479b-a95d-703e62328958)

ad-hoc calc setting with lots of tab-moving

**Comment by caolanm** [comment]
This weeks calc profile

![perf-17266](https://github.com/CollaboraOnline/online/assets/833656/6b5f5e81-4ed5-4612-a783-b4441d1b274b)

**Comment by caolanm** [comment]
![perf-209009](https://github.com/CollaboraOnline/online/assets/833656/ae089a55-a590-4a9f-b4a4-43f0b161bcfd)

todays profile, I imagine the calendar swapping is not typical workload

**Comment by caolanm** [comment]
I'll put calc 24.04 flamegraphs at: https://github.com/CollaboraOnline/online/issues/8570

**Comment by caolanm** [comment]
![perf-581706](https://github.com/CollaboraOnline/online/assets/833656/1ed02167-eeb2-4481-89df-87c9c1a18de3)

Todays 23.05 flamegraph. There looks to be a lot of chart updating, which fits with what we saw visually wrt auto-updating chart from random numbers. And there is a lot of calendar switching, (Gregorian to Hebrew?) which I imagine is not normal workload.

**Comment by mmeeks** [comment]
20% of the time in ScTable::FindMaxRotCol - is not a great score ;-) ( aggregated across many different callers manually - 4.8%+5.2%+5.5%+1.3%+3.3% etc.)

**Comment by caolanm** [comment]
![perf-302955](https://github.com/CollaboraOnline/online/assets/833656/bed3e054-9b41-4517-b85a-04982342e82e)

calc 23.05 profile ah hoc testing

**Comment by caolanm** [comment]
![perf-304010](https://github.com/CollaboraOnline/online/assets/833656/50915738-0aad-403e-b4fc-20b68d7f9d14)

2nd half of calc 23.05 session

**Comment by caolanm** [comment]
![perf-308642](https://github.com/CollaboraOnline/online/assets/833656/b40597f5-d584-4fce-b8d7-f34ee71424a5)

and follow up calc 23.05 test without copy and paste which was more performant

**Comment by zefilipe** [comment]
Hi,

I Recently updated LibreOfficeto version:

 Version: 24.8.1.2 (X86_64) / LibreOffice Community
Build ID: 87fa9aec1a63e70835390b81c40bb8993f1d4ff6
CPU threads: 8; OS: Windows 10 X86_64 (10.0 build 19045); UI render: Skia/Raster; VCL: win
Locale: pt-PT (pt_PT); UI: en-US
Calc: threaded

And when i press CTRL+S to save, sometimes its crash, open a window to send or not send crash report, and i can use safe mode or not, and repair the file ... !

In previows version this dont happend !

Regards

**Comment by caolanm** [comment]
![perf-2575](https://github.com/user-attachments/assets/e0e05f2c-c667-4a0f-9ff4-ce9f1d8c6ceb)

Today's COOL meeting calc testing session, odd to see std::chrono::steady_clock::now() in Document::trimAfterInactivity featuring

**Comment by caolanm** [comment]
We created a separate task for 25.04 https://github.com/CollaboraOnline/online/issues/11336 so lets close this one now
